dist/opt/flight/libexec/flight-starter/main.sh

```diff
@@ -30,6 +30,14 @@ if [ "$1" == "start" ]; then
   if [ -z "${flight_DEFINES}" ]; then
     flight_DEFINES=(flight_ACTIVE)
     flight_DEFINES_exits=()
+    flight_DEFINES_paths=" "
+    # NOTE: The "export flight_DEFINES" and "export flight_DEFINES_exists"
+    # does not work as bash cannot export arrays.
+    #
+    # The broken exports are being maintained for reference, one of the
+    # following should be done:
+    # * Update the relevant calls to use a string, or
+    # * Remove the exports
```

Comment:
It would seem that the `export` line has lead you to misunderstand how this works.

When the Flight Environment is started (`flight start`) the environment of the *current* shell is modified.  The modification consists of a number of functions being defined, new environment variables being set and existing environment variables being modified.

As part of the Flight Environment being started the profile scripts are `source`d.  They also modify the environment of the *current* shell.

When a Flight Env environment is activated (`flight env activate ...`), funky things happen to cause the `activate` script, e.g., `/opt/flight/usr/lib/env/types/conda/activate.bash.erb` to be `source`d into the *current* shell.  The activate script further modifies the environment of the *current* shell.

As all of these modifications are being performed in the *current* shell, the environment variables `flight_DEFINES` and `flight_DEFINES_exits` are readily accessible as arrays.  They do not need to be accessible to subshells of subprocesses as they cannot modify the environment of the *current* shell.

That leads us to the question of "why are the exported"?  To which I don't have a good answer.  Its probably not required.

---

The `flight_DEFINES` and `flight_DEFINES_exits` process does work and doesn't need modification as can be demonstrated with the following.

Below I modify the `PATH` and define an exit function to remove that modification (the function is obviously not robust).

```
[vagrant@chead1 ~]$ declare | grep -E '^flight_DEFINES'
[vagrant@chead1 ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/vagrant/.local/bin:/home/vagrant/bin
[vagrant@chead1 ~]$ 
[vagrant@chead1 ~]$ flight start
<SNIP>
OpenFlight is now active.
[vagrant@chead1 (dev) ~]$ 
[vagrant@chead1 (dev) ~]$ declare | grep -E '^flight_DEFINES'
flight_DEFINES=([0]="flight_ACTIVE" [1]="flstop" [2]="flight_stop" [3]="flight" [4]="fl" [5]="flexec" [6]="flactivate" [7]="flintegrate" [8]="flenv" [9]="flight_env" [10]="flight_env_exit" [11]="flight_ENV_root" [12]="flight_ENV_shell")
flight_DEFINES_exits=([0]="flight_env_exit")
flight_DEFINES_orig_PS1='[\u@\h \W]\$ '
[vagrant@chead1 (dev) ~]$ 
[vagrant@chead1 (dev) ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/vagrant/.local/bin:/home/vagrant/bin:/opt/flight/opt/slurm/bin
[vagrant@chead1 (dev) ~]$ 
[vagrant@chead1 (dev) ~]$ PATH=$PATH:/MY/BIN/DIR
[vagrant@chead1 (dev) ~]$ remove_my_bin_dir() { echo "Removing my bin dir"; PATH=$(echo "$PATH" | sed 's,:/MY/BIN/DIR,,g') ; }
[vagrant@chead1 (dev) ~]$ flight_DEFINES_exits+=(remove_my_bin_dir)
[vagrant@chead1 (dev) ~]$ 
[vagrant@chead1 (dev) ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/vagrant/.local/bin:/home/vagrant/bin:/opt/flight/opt/slurm/bin:/MY/BIN/DIR
[vagrant@chead1 (dev) ~]$ 
[vagrant@chead1 (dev) ~]$ flight stop
Removing my bin dir
OpenFlight is now inactive.
[vagrant@chead1 ~]$ echo $PATH
/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/vagrant/.local/bin:/home/vagrant/bin:/opt/flight/opt/slurm/bin
```